substrate/tendermint/client/src/tendermint.rs

```diff
@@ -0,0 +1,228 @@
+  pub(crate) importing_block: Arc<RwLock<Option<<T::Block as Block>::Hash>>>,
+  pub(crate) recheck: Arc<RwLock<HashSet<<T::Block as Block>::Hash>>>,
+
+  pub(crate) client: Arc<T::Client>,
+  pub(crate) queue:
+    Arc<AsyncRwLock<Option<TendermintImportQueue<T::Block, T::BackendTransaction>>>>,
+}
+
+impl<T: TendermintValidator> Clone for TendermintImport<T> {
+  fn clone(&self) -> Self {
+    TendermintImport {
+      validators: self.validators.clone(),
+
+      providers: self.providers.clone(),
+      importing_block: self.importing_block.clone(),
+      recheck: self.recheck.clone(),
+
+      client: self.client.clone(),
+      queue: self.queue.clone(),
+    }
+  }
+}
+
+impl<T: TendermintValidator> TendermintImport<T> {
+  pub(crate) fn new(client: Arc<T::Client>) -> TendermintImport<T> {
+    TendermintImport {
+      validators: TendermintValidators::new(client.clone()),
+
+      providers: Arc::new(AsyncRwLock::new(None)),
+      importing_block: Arc::new(RwLock::new(None)),
+      recheck: Arc::new(RwLock::new(HashSet::new())),
+
+      client,
+      queue: Arc::new(AsyncRwLock::new(None)),
+    }
+  }
+
+  pub(crate) async fn inherent_data(&self, parent: <T::Block as Block>::Hash) -> InherentData {
+    match self
+      .providers
+      .read()
+      .await
+      .as_ref()
+      .unwrap()
+      .create_inherent_data_providers(parent, ())
+      .await
+    {
+      Ok(providers) => match providers.create_inherent_data() {
+        Ok(data) => Some(data),
+        Err(err) => {
+          warn!(target: "tendermint", "Failed to create inherent data: {}", err);
+          None
+        }
+      },
+      Err(err) => {
+        warn!(target: "tendermint", "Failed to create inherent data providers: {}", err);
+        None
+      }
+    }
+    .unwrap_or_else(InherentData::new)
+  }
+
+  async fn check_inherents(
+    &mut self,
+    hash: <T::Block as Block>::Hash,
+    block: T::Block,
+  ) -> Result<(), Error> {
+    let inherent_data = self.inherent_data(*block.header().parent_hash()).await;
+    let err = self
+      .client
+      .runtime_api()
+      .check_inherents(&BlockId::Hash(self.client.info().finalized_hash), block, inherent_data)
+      .map_err(|_| Error::Other(BlockError::Fatal.into()))?;
+
+    if err.ok() {
+      self.recheck.write().unwrap().remove(&hash);
+      Ok(())
+    } else if err.fatal_error() {
+      Err(Error::Other(BlockError::Fatal.into()))
+    } else {
+      self.recheck.write().unwrap().insert(hash);
+      Err(Error::Other(BlockError::Temporal.into()))
+    }
+  }
+
+  // Ensure this is part of a sequential import
+  pub(crate) fn verify_order(
+    &self,
+    parent: <T::Block as Block>::Hash,
+    number: <<T::Block as Block>::Header as Header>::Number,
+  ) -> Result<(), Error> {
+    let info = self.client.info();
+    if (info.finalized_hash != parent) || ((info.finalized_number + 1u16.into()) != number) {
+      Err(Error::Other("non-sequential import".into()))?;
```

Comment:
it would be helpful to print out the parent and number

Reply:
Also adding debugs to all slash conditions.